es6/animations.js

```diff
@@ -5,180 +5,176 @@ import {
-        start_time = end_time - duration
-    if (end_time === undefined)
-        end_time = start_time + duration
-    if (duration === undefined)
-        duration = end_time - start_time
+export const Become = ({path, state, start_time,
+                        end_time=Infinity, duration=Infinity}) => {
+    if (start_time === undefined) start_time = Date.now()
 
-    if (start_time + duration != end_time) {
-        console.log({start_time, end_time, duration})
-        throw 'Conflicting values, Animation end_time != start_time + duration'
+    if (exactlyOneIsUndefined(duration, end_time)) {
+        [duration,
+             end_time] = computeTheOther(start_time, duration, end_time)
+    }
+    if (start_time === undefined || path === undefined) {
+        // console.log({path, state, start_time, end_time, duration})
+        throw 'Become animation must have a start_time and path defined.'
     }
-    return {start_time, duration, end_time}
+    return Animate({
+        type: 'BECOME',
+        path,
+        start_state: state,
+        end_state: state,
+        start_time,
+        end_time,
+        duration,
+        tick: (time_elapsed) => {
+            return state
+        },
+    })
 }
 
-
-const checked_animation_amt = ({key, start_state, end_state, amt}) => {
-    if (typeof(start_state) === 'number') {
-        if ([start_state, end_state, amt].filter(a => typeof(a) == 'number').length < 2) {
-            console.log({start_state, end_state, amt})
-            throw 'Need at least 2/3 to calculate animation: start_state, end_state, amt'
-        }
-
-        if (start_state === undefined)
-            start_state = end_state - amt
-        if (end_state === undefined)
-            end_state = start_state + amt
-        if (amt === undefined)
-            amt = end_state - start_state
-
-        if (start_state + amt != end_state) {
-            console.log({start_state, end_state, amt})
-            throw 'Conflicting values, Animation end_state != start + amt'
+export const computeTheOther = (start, delta, end) => {
+    if (typeof(start) === 'object') {
+        let new_delta = delta ? {...delta} : {}
+        let new_end = end ? {...end} : {}
+        Object.keys(start).forEach((key) => {
+            let [_delta, _end] = computeTheOther(start[key],
+                                                 new_delta[key],
+                                                 new_end[key])
+            new_delta[key] = _delta
+            new_end[key] = _end
+        })
+        return [new_delta, new_end]
+    }
+    if (typeof(start) === 'number') {
+        if (end === undefined && delta !== undefined) {
+            return [delta, start + delta]
+        } else if (end !== undefined && delta === undefined) {
+            return [end - start, end]
+        } else if (start + delta !== end) {
+            console.log((new Error()).stack)
+            throw `start (${start}) + delta (${delta}) !== end (${end})`
```

Comment:
What happened to:

```javascript
throw 'Incompatible types passed as {start_state, end_state, amt}, must all be objects or numbers'
```

Reply:
Because that's an arbitrary restriction that's not even followed in your test cases. Why can't an animation set state to a string, e.g.?